### ISSUE TITLE
callCabal2nix is broken on nixos-20.09

### ISSUE THREAD

**Issue by bgamari** [NixOS/nixpkgs#108246]
**Describe the bug**

Evaluation of the follow expression:

```nix
{ baseNixpkgs ? (import <nixpkgs> {}) }:

let
  nixpkgsSrc = baseNixpkgs.fetchFromGitHub {
    owner = "nixos";
    repo = "nixpkgs";
    rev = "ffb3aab257e8851b558cdc6079241a7eb0c7239e";
    sha256 = "sha256:147d4nsq0334skyr9rghcfnizasvba1pr2rq72axwdldlma0qli1";
  };
  nixpkgs = import nixpkgsSrc {};
in

with nixpkgs;
let
  src = fetchFromGitHub {
    owner = "haskell";
    repo = "Cabal";
    rev = "cabal-install-3.4.0.0-rc3";
    sha256 = "1m1355863s4cj9v4r28b6d4kkhh3m7h6gjn1ziffl76r0xpk94rh";
  };
in
haskellPackages.callCabal2nix "cabal-install" src {}
```
fails with:
```
builder for '/nix/store/s6dcvlwxbbah99k4kmwcmngrxwimgy2l-cabal2nix-cabal-install.drv' failed with exit code 1; last 2 log lines:
  installing
  cabal2nix: nix-prefetch-url: createProcess: runInteractiveProcess: exec: does not exist (No such file or directory)
[0 built (1 failed)]
error: build of '/nix/store/s6dcvlwxbbah99k4kmwcmngrxwimgy2l-cabal2nix-cabal-install.drv' failed
```

**To Reproduce**
1. See above

**Expected behavior**

Evaluation doesn't fail with the above error.

**Screenshots**

**Additional context**

**Notify maintainers**

@peti 

**Metadata**
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.9.11, NixOS, 20.09.2076.ffb3aab257e (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.9`
 - channels(root): `"nixos-20.09.2076.ffb3aab257e, nixpkgs-20.09.2076.ffb3aab257e"`
 - channels(ben): `"nixpkgs-20.09.2076.ffb3aab257e, nixpkgs-20_03-20.03.3310.9518fac712c, nixpkgs-unstable-21.03pre253752.272744825d2, nixos-18.09-18.09.2574.a7e559a5504"`
 - nixpkgs: `/home/ben/.nix-defexpr/channels/nixpkgs`

```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by bgamari** [comment]
@srk points out that `haskellSrc2nix` has `nativeBuildInputs = [ buildPackages.cabal2nix-unwrapped ];`. This seems  wrong and is likely responsible for this failure. Afterall, the `cabal2nix` wrapper on my machine looks like:
```bash
#! /nix/store/2jysm3dfsgby5sw5jgj43qjrb5v79ms9-bash-4.4-p23/bin/bash -e
export PATH='/nix/store/xhz3ybvfn5s1qiza4fijd2l05cfd6082-nix-2.3.9/bin:/nix/store/vgfsj965y0gv2z09v1yjay75m6ac94z4-nix-prefetch-scripts/bin'${PATH:+':'}$PATH
exec -a "$0" "/nix/store/pxzllfyy4qiblxv6dibfmpg34b7jdbi4-cabal2nix-2.15.5/bin/.cabal2nix-wrapped"  "$@"
```
Which appears to ensure that `nix-prefetch-url` is in `PATH`.

**Comment by cdepillabout** [comment]
@bgamari Sorry you're having trouble with this.

The unwrapped version of `callCabal2nix` was introduced in https://github.com/NixOS/nixpkgs/pull/81125 (cc @madjar).

> ```
> builder for '/nix/store/s6dcvlwxbbah99k4kmwcmngrxwimgy2l-cabal2nix-cabal-install.drv' failed with exit code 1; last 2 log lines:
>   installing
>   cabal2nix: nix-prefetch-url: createProcess: runInteractiveProcess: exec: does not exist (No such file or directory)
> ```

It seems somewhat strange that you'd be getting this error, since you already pulled down the source with `fetchFromGithub`.

Do you have any idea why `cabal2nix` would need to be fetching something else with `nix-prefetch-url`?

**Comment by bgamari** [comment]
> Do you have any idea why cabal2nix would need to be fetching something else with nix-prefetch-url?

I am afraid that I haven't the slightest idea.

**Comment by roberth** [comment]
I've seen this error message when something else was wrong. Cabal2nix seems to fall back to other fetching methods in some cases. That won't work in the sandbox of course. It should only try to use the local package and just report what's wrong. I don't remember the exact error condition, but probably something like a missing cabal file, wrong path. That kind of thing.

**Comment by peti** [comment]
`cabal2nix` should not do this. It would use `nix-prefetch-url` to download a release tarball from Hackage, for example, to figure out the SHA512 of that tarball to put into the generated expression. Since the source code is found *locally*, however, it should not need a hash and consequently shouldn't need to download anything. You can probably work around this issue by using `callCabal2nixWithOptions` and passing an extra `--sha256=deadbeef` argument to make sure it doesn't try to guess that hash. It's just a wild guess, though, I'm not sure what is going on either.

**Comment by cdepillabout** [comment]
@sternenseemann Could this be related to the problem you're recent been looking into?  For example, in https://github.com/NixOS/cabal2nix/pull/489?

**Comment by sternenseemann** [comment]
Yes, this should be precisely the issue that the new error message benefits from. However, I can't reproduce the original issue with current `haskell-updates` which isn't using the unstable version of cabal2nix (that has the fix) yet. Seems like something else changed that resolved the issue.

**Comment by jmininger** [comment]
It doesn't look like the same cause, but I figured that it might be useful to add that I ran into the same exact problem when i mispelled the name of "my-package" in `haskellPackages.callCabal2nix "my-package" ./. {}`
